src/coroutine.js

```diff
@@ -25,6 +26,36 @@ export default function coroutine (action, body, repo) {
     return action
   }
 
+  if (body && get(body, 'constructor.name') === 'GeneratorFunction') {
+    action.open()
+
+    let iterator = body(repo.push.bind(repo), repo)
+
+    function step (payload) {
+      let next = iterator.next(payload)
+
+      if (next.done) {
+        action.resolve(payload)
+      } else {
+        progress(next.value)
+      }
+    }
+
+    function progress (value) {
+      if (value instanceof Action) {
+        value.onDone(step)
+        value.onCancel(action.cancel, action)
+        value.onError(action.reject, action)
+      } else {
+        step(value)
+      }
```

Comment:
I think asserting that an action has been passed in would be nice here, you could make the functionality of this a little more descriptive as well:
```js
function progress (subAction) {
  subAction.onDone(step)
  subAction.onCancel(action.cancel, action)
  subAction.onError(action.reject, action)
}